### PR TITLE
[REVIEW] Enable strict_decimal_types in parquet reading

### PR DESCRIPTION
**Pull request by sperlingxx**
This pull request is to address #6909.

### CODE REVIEW

**Review by revans2 — APPROVED** [review]
The java side changes are just a comment change and look fine to me.

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:825]: Does this really need to be volatile?

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:829]:
```suggestion
  const uint8_t *dict = s->dict_base ? s->dict_base : s->data_start;
```

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:830]:
```suggestion
  uint32_t dict_pos = (s->dict_base ? ((s->dict_bits > 0) ? s->dict_idx[src_pos & (non_zero_buffer_size - 1)] : 0) : src_pos) * dtype_len_in;
  uint32_t dict_size = s->dict_size;
  uint32_t dtype_len_in = s->dtype_len_in;
```

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:841]:
```suggestion

```

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:842]:
```suggestion
```

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:1732]: Why would we only want to do this with strict decimal types? That flag means that we blow up if we have to coerce the value into a double. I think we would want this all the time and read decimals as decimals whenever possible.

File: cpp/src/io/parquet/reader_impl.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/reader_impl.cu:148]:
```suggestion
      else if (physical == parquet::FIXED_LEN_BYTE_ARRAY && schema.type_length <=8) {
        return type_id::DECIMAL64;
      } else {
        CUDF_EXPECTS(strict_decimal_types == false, "Unsupported decimal type read!");
        return type_id::FLOAT64;
      }
```

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:1732]:
```suggestion
              // we currently do not support reading byte arrays larger than DECIMAL64
              if (dtype_len <= 8) {
```

File: cpp/src/io/parquet/parquet_gpu.hpp
Comment (hyperbolic2346) [inline cpp/src/io/parquet/parquet_gpu.hpp:210]: I think most of this can go away if we take the suggestion to do this all the time.

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
Great change! I like adding more features here and reading byte arrays that are small enough. This will allow decimal support for more data. Thanks!

**Comment by harrism** [comment]
@sperlingxx please allow the automatic reviewer assignment to do its job. I don't think my review is required here?

File: cpp/src/io/parquet/parquet_gpu.hpp
Comment (sperlingxx) [reply cpp/src/io/parquet/parquet_gpu.hpp]: Yes, I've removed it.

File: cpp/src/io/parquet/page_data.cu
Comment (sperlingxx) [reply cpp/src/io/parquet/page_data.cu:825]: I'm not sure.I added volatile here because all other device decoding function regard `page_state_s` as a volatile variable.

File: cpp/src/io/parquet/page_data.cu
Comment (sperlingxx) [reply cpp/src/io/parquet/page_data.cu]: Yes, you are right. I misunderstood the meaning of `strict_decimal_types`. Thanks for correcting me.

**Comment by sperlingxx** [comment]
> @sperlingxx please allow the automatic reviewer assignment to do its job. I don't think my review is required here?

I'm sorry for that. I invited you because it seems to require the review from `cudf-cpp-codeowners`.

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [reply cpp/src/io/parquet/page_data.cu:825]: Fair enough. :)

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:843]: We should comment this a little. Something like // If the value isn't all 8 bytes we need to shift it up and back down in order to sign extend the value so a negative number is correctly represented. Someone who doesn't know this trick will appreciate the comment.

File: cpp/src/io/parquet/page_data.cu
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:1727]: Should we do the same thing fir 32-bit values so we can store them as compactly as possible?

Comment (hyperbolic2346) [reply cpp/src/io/parquet/page_data.cu:1727]: It is perfectly fine to say "yes, in a different PR" 😄

File: cpp/src/io/parquet/page_data.cu
Comment (sperlingxx) [reply cpp/src/io/parquet/page_data.cu:845]: Good point! It's done.

File: cpp/src/io/parquet/page_data.cu
Comment (sperlingxx) [reply cpp/src/io/parquet/page_data.cu:1729]: I am not sure whether it is necessary. In terms of spark-rapids, it only supports `DECIMAL64` (`INT64`) in current.

**Review by hyperbolic2346 — APPROVED** [review]
Looks good, thanks for doing this!

**Comment by sperlingxx** [comment]
@nvdbaranec Could you help to review this pull request? I think we need at least one approval from `cudf-cpp-codeowners`.

File: cpp/src/io/parquet/page_data.cu
Comment (devavret) [inline cpp/src/io/parquet/page_data.cu:831]: A bit questionable naming of variables here. It appears that depending on the encoding, the `dict` ptr points to either the dict or the data. Similarly for the variable `dict_pos`. Could've been `data` and `pos`. But this seems to be the trend in the entire file.

Comment (sperlingxx) [reply cpp/src/io/parquet/page_data.cu:831]: Yes, @devavret. Comparing to `dict` and `dict_pos`,  `data` and `pos` are better names.  The name style here is just following the taste of other decoding methods. Do you think we need to change to  `data` and `pos` ?

Comment (devavret) [reply cpp/src/io/parquet/page_data.cu:831]: This is just something we'll need to remember during the refactor.

**Review by devavret — APPROVED** [review]

File: cpp/src/io/parquet/page_data.cu
Comment (devavret) [reply cpp/src/io/parquet/page_data.cu:831]: Yes please. Just change the ones in this function. Some functions are doing this correctly https://github.com/rapidsai/cudf/blob/f117b68ca68faa0acd8f9fa485cad0aa63a0e543/cpp%2Fsrc%2Fio%2Fparquet%2Fpage_data.cu#L627-L629

**Pull request merged**
